### PR TITLE
gwctl: Decouple data model from printing for flexibility and clarity

### PR DESCRIPTION
**Pull request by gauravkghildiyal**
## What type of PR is this?
/kind cleanup

## What this PR does / why we need it:
This PR significantly improves the organization and flexibility of our resource model, making it easier to maintain and extend.

Key highlights:
- Modular design: The data model is now independent of the printing view, simplifying maintenance, testing and future enhancements.
- Enhanced extensibility:
    - This separation paves the way for quickly implementing detailed printed describe views with richer information. 
    - The graph-based model may potentially enable visualizations using tools like Graphviz or interactive web views, offering a valuable way to explore resource relationships in the future.
- Improved code clarity: The new structure promotes better code understanding, encouraging contributions.

## Which issue(s) this PR fixes:

Fixes #

## Does this PR introduce a user-facing change:

```release-note
NONE
```

### CODE REVIEW

**Review by robscott — COMMENTED** [review]
Thanks for the work on this @gauravkghildiyal! I added some comments/questions, but also have a broader overall thought: This code is getting sufficiently complex that it could benefit from some kind of high level description of the architecture, maybe just in the gwctl README? I think something like that could help others pick this up and contribute more quickly.

File: gwctl/pkg/resourcediscovery/resourcemodel.go [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+
+// ResourceModel represents the graph-like model of Gateway API resources and
+// their relationships, capturing dependencies and interactions. It acts as a
+// central data structure for organizing and managing these resources, enabling
+// operations like:
+//   - Discovering and understanding the relationships between different
+//     resources
+//   - Calculating effective policies based on hierarchical inheritance
+//   - Identifying potential conflicts or issues in resource configuration
+//   - Visualizing the topology of Gateway API resources
```
[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply gwctl/pkg/resourcediscovery/resourcemodel.go:39]: That's right, Rob! I've added a `CODE_ORGANIZATION.md` file covering how individual packages/structures connect with each other. Have intentionally tried to keep it simple.

**Review by gauravkghildiyal — COMMENTED** [review]
Thanks for the review! 
I've made the changes and added a doc which has a high level description of the various pieces. I agree that would help others pick things up.

File: gwctl/CODE_ORGANIZATION.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                     +---------->| HTTPRoute |             | HTTPRoute |            | HTTPRoute |
+                                 +----+------+             +----+------+            +-----+-----+
+                                      |                         |                         |
+                                      |                         |                         |
+                                  Routes to                 Routes to                 Routes to
+                                      |                         |                         |
+                                      |       +---------+       |                    +----v----+
+                                      +-------> Backend <-------+                    | Backend |
+                                              +---------+                            +---------+
+            ```
```
Comment (robscott) [inline gwctl/CODE_ORGANIZATION.md:58]: Non-blocking - this _might_ be easier to represent with a [Mermaid.js diagram](https://mermaid.js.org/).

**Review by robscott — COMMENTED** [review]
Thanks @gauravkghildiyal!

/lgtm
/approve

**Pull request merged**